hpcrawler.py

```diff
@@ -0,0 +1,364 @@
+#           compare it to the ones in <p>, and we will update the corresponding track in <p>
+#           to include the new features. if a track isn't on <p> , we'll mine it all up.
+#       (3) create different:
+#               * directory <tracks_dir_path>: we don't need to overwrite the whole data set.
+#               * file <p>
+#           so maybe we should give them as parameters to the class?
+#
+
+
+class HpCrawler:
+    """
+    This class crawls "The Hiking Project"'s website: https://www.hikingproject.com/.
+    It process the data crawled, and saves it, as follows:
+    it creates a dir named "hp", that contains 2 dirs and a file:
+     (1) dir <gpx_dir_path> : contains country directories, each directory holds all gpx files of the tracks
+                            crawled from that country and a json file keeping trace of the data collecting progress
+                            * The dir is named by the country name, string: <country>
+                            * The gpx files are named by the track's idx, int >= 0: <j>
+                            * The json file is named <progress>: maps [track_gpx_filename, track_length,
+                                track_difficulty] to track idx <j>, int >=0.
+     (2) dir <tracks_dir_path> : contains json files. These files are named by the length_tag of the tracks
+                                they hold: int >= 0 denoted as <l>.
+                                each json files holds a dict mapping the track name: <country><j>
+                                to the track's representation.
+     (3) json file <seen_path> : Keeps trace of the mining progress: contains a dict that maps the progress
+                                to the country <country>:
+                                [idx_of_next_track_we_need_to_crawl, [<urls>]] if we're mid process,
+                                and 'Done' otherwise.
+    """
+
+    # static fields:
+    wait = 5
+    done_tag = 'DONE'
+    gpx_dir_path = 'hp\\gpx'
+    tracks_dir_path = 'hp\\tracks'
+    seen_path = 'hp\\seen.json'
+
+    def __init__(self, to_crawl):
+        """
+        crawls the trails data from "the hiking project"'s site.
+        it saves the data and the progress of crawling task under the "hp" directory.
+        :param to_crawl: python list of countries to crawl.
+                        the countries should start with a capital letter.
+        """
+
+        self._countries = to_crawl
+        self._country = None  # string
+        self._track_idx = str(0)  # always string
+        self._path = None  # the location of collected data: gpx and progress of collecting (not final data)
+        self._url = None  # python list of strings
+        self._driver = None  # firefox(!) driver
+        self._driver_status = False
+
+    def __del__(self):
+        """
+        closes the driver in case an exception that hasn't been treated has occurred
+        """
+        if self._driver_status is True:
```

Comment:
would be easier to call _setup in __init__ and totally remove _driver_status variable